File: 543 Goldbach's Conjecture/solution.c
#include <stdio.h>
#include <stdlib.h>
#define MAX_NUM 999995
#define GET(n) (mark[(n) >> 5] & (1u << ((n) & 31)))
#define SET(n) (mark[(n) >> 5] |= (1u << ((n) & 31)))

unsigned int mark[(MAX_NUM >> 6) + 1];
int *primes, size;

void eulerSieve() {
    for(int n = 3; n <= MAX_NUM; n += 2) {
        if(!GET(n >> 1)) {
            primes = (int*)realloc(primes, (size + 1) * sizeof(int));
            primes[size++] = n;
        }
        for(int i = 0, temp; (temp = primes[i] * n) <= MAX_NUM; ++i) {
            SET(temp >> 1);
            if(n % primes[i] == 0)
                break;
        }
    }
}

int main() {
    eulerSieve();
    int n;
    while(scanf("%d", &n) && n != 0) {
        int i = 0;
        while(GET(n - primes[i] >> 1))
            ++i;
        printf("%d = %d + %d\n", n, primes[i], n - primes[i]);
    }
    free(primes);
	return 0;
}

File: 543 Goldbach's Conjecture/solution.java
import java.io.*;
import java.util.*;

public class Main {
    static int MAX_NUM = 999995;
    static boolean[] isComposite = new boolean[(MAX_NUM >> 1) + 1];
    static ArrayList<Integer> primes = new ArrayList<>();
	public static void main(String[] args) throws IOException {
    	eulerSieve();
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
    	int n;
    	StringBuilder output = new StringBuilder();
    	while(st.nextToken() == StreamTokenizer.TT_NUMBER && (n = (int)st.nval) != 0) {
	        int i = 0;
	        while(isComposite[n - primes.get(i) >> 1])
	            ++i;
	        output.append(n).append(" = ").append(primes.get(i)).append(" + ").append(n - primes.get(i)).append('\n');
    	}
	    System.out.print(output);
	}
	
	static void eulerSieve() {
        for(int n = 3; n <= MAX_NUM; n += 2) {
            if(!isComposite[n >> 1])
                primes.add(n);
            for(int i = 0, temp; (temp = primes.get(i) * n) <= MAX_NUM; ++i) {
                isComposite[temp >> 1] = true;
                if(n % primes.get(i) == 0)
                    break;
            }
        }
    }
}
